spock/src/main/java/org/jboss/weld/spock/impl/EnableWeldInterceptor.java

```diff
@@ -0,0 +1,187 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jboss.weld.spock.impl;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Parameter;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.WeakHashMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import jakarta.enterprise.inject.spi.BeanManager;
+import org.jboss.weld.inject.WeldInstance;
+import org.jboss.weld.spock.WeldInitiator;
+import org.jboss.weld.spock.WeldSpockEnricher;
+import org.spockframework.runtime.extension.IMethodInterceptor;
+import org.spockframework.runtime.extension.IMethodInvocation;
+import org.spockframework.runtime.model.FeatureInfo;
+
+import static java.util.Collections.singleton;
+import static java.util.Collections.synchronizedMap;
+import static java.util.stream.Collectors.toList;
+import static org.spockframework.runtime.model.MethodInfo.MISSING_ARGUMENT;
+
+/**
+ * A Spock interceptor, that serves as base for the manual and automatic interceptors and has the common logic.
+ * It first calls the custom logic of the subclass, then adds the shared instance as injection target, initializes
+ * the Weld container and after invocation proceeded, shuts down the container. It also provides methods to get a
+ * test instance injector and a parameter injector as these need to be attached to varying extension points,
+ * depending on the annotation configuration.
+ *
+ * @author Björn Kautler
+ */
+abstract class EnableWeldInterceptor implements IMethodInterceptor {
+    protected final List<WeldSpockEnricher> weldSpockEnrichers;
+    private final Map<IdentityWrapper, WeldInitiator> weldInitiators = synchronizedMap(new WeakHashMap<>());
+    private final List<FeatureInfo> handledFeatures = new CopyOnWriteArrayList<>();
+
+    public EnableWeldInterceptor(List<WeldSpockEnricher> weldSpockEnrichers) {
+        this.weldSpockEnrichers = weldSpockEnrichers;
+    }
+
+    protected abstract WeldInitiator weldInit(IMethodInvocation invocation);
+
+    @Override
+    public void intercept(IMethodInvocation invocation) throws Throwable {
+        WeldInitiator weldInitiator = weldInit(invocation);
+        weldInitiator.addObjectsToInjectInto(singleton(invocation.getSharedInstance()));
+        weldInitiator.initWeld(invocation.getInstance());
+        try {
+            IdentityWrapper id = (invocation.getIteration() == null) ? null : new IdentityWrapper(invocation.getInstance());
+            weldInitiators.put(id, weldInitiator);
```

Comment:
Line 114 explains it shortly.

You have basically 4 situations.
a: a specification scoped container and interceptor that is started before the first feature (test) in a specification (test class) and shut down after the last feature in a specification
b: a feature scoped container and interceptor that is started before the feature and is shut down after the feature where there are multiple iterations, so there is a `where:` block that runs the feature repeatedly with different inputs (i. e. data-driven tests)
c: like b, but without a `where:` block and thus with only one iteration
d: an iteration scoped container and interceptor that is stated before an iteration and shut down after the iteration (the default)

For a and b when the main interceptor runs, the test instance is not yet available and the initiator is added with `null` key and subsequently used for all iterations that run underneath it.
c is a bit uninteresting, it has feature scope, but there is only one iteration, the iteration is already available when the main interceptor runs but there will only be one initiator like for a and b.
For d you have also just one main interceptor for one feature, but you want an own initiator for each iteration, so in case of a data-driven feature now you fill the map with one initiator per test instance as key.

Reply:
Alternatively I could probably use the shared instance as key.
The shared instance is an additional instance of the specification where in Spock the `@Shared` fields are kept, they are much like `static` fields, but have some advantages and should almost always be preferred.
But I thought `null` is cleaner there.

Reply:
Ok, so `null` is just a 'marker value' for that special case, got it.
Thanks for explanation, it now makes sense as-is.